File: coolify/apps/nextjs-kiosk/src/lib/device-profile.ts
/**
 * device-profile.ts — Android ラッパーが署名して送る端末プロファイル。
 *
 * ラッパー（KioskBridge.deviceProfile(nonce)）が、TOFU 束縛済みの Keystore 鍵で
 * `nonce\nprofileJson` に署名して返す。サーバーは**署名を検証してから**
 * profileJson を parse する（順序が逆だと、未検証の JSON をパースする穴になる）。
 *
 * ■ 何が保証されるか
 * 鍵は端末の非エクスポート Keystore にあるので、プロファイルを 1 文字でも
 * 書き換えれば署名が壊れる = **その端末が申告した内容である**ことは保証できる。
 *
 * ■ 何が保証されないか
 * root 化した端末が「本物の鍵で嘘の値に署名する」ことは防げない。そこを塞ぐのは
 * ハードウェア鍵アテステーション（setAttestationChallenge + Google ルートへの
 * チェーン検証）だけで、既存の TOFU 束縛を作り直すことになるため今回は採らない。
 *
 * ■ nonce の二重確認
 * プロファイル内にも nonce を入れさせ、外側の nonce と一致することを見る。
 * 別チャレンジで得た署名済みプロファイルの貼り替えを防ぐ。
 */

// server-only は付けない — attest-core.ts と同じく DB も Cookie も触らない
// 純粋な検証ロジックで、そのままユニットテストから呼びたいため。
import { z } from "zod";
import { verifyDeviceSignature } from "@/lib/attest-core";
import type { VerifiedWrapperProfile } from "@/lib/device-ownership-core";

/** プロファイルの版。ラッパー側 KioskBridge と合わせる。 */
export const PROFILE_SCHEMA_VERSION = 1;

/** 署名対象文字列の**唯一の定義**。ラッパー側と 1 文字も違ってはいけない。 */
export function attestPayload(
  nonce: string,
  profileJson: string | null,
): string {
  return profileJson === null ? nonce : `${nonce}\n${profileJson}`;
}

/**
 * 端末プロファイルのスキーマ。**すべて任意**にしてあるのは、minSdk 29 では
 * enrollmentId（API31+）や installer（API30+）が普通に取れないため。
 * サーバー側は null 前提で書くこと。
 */
export const deviceProfileSchema = z.object({
  v: z.number().int(),
  nonce: z.string().min(1),
  signedAt: z.number().int().optional(),

  appVersion: z.string().max(32).optional(),
  appVersionCode: z.number().int().optional(),
  packageName: z.string().max(128).optional(),
  installer: z.string().max(128).nullable().optional(),

  // 管理状態（所有区分の判定材料）
  isDeviceOwner: z.boolean().optional(),
  isProfileOwner: z.boolean().optional(),
  isManagedProfile: z.boolean().optional(),
  activeAdmins: z.number().int().optional(),
  lockTaskState: z.number().int().optional(),
  /** dpm.getEnrollmentSpecificId() — 組織×端末×アプリで一意・初期化しても不変 */
  enrollmentId: z.string().max(128).nullable().optional(),

  // 端末同定
  androidId: z.string().max(64).nullable().optional(),
  serial: z.string().max(64).nullable().optional(),

  // Build
  manufacturer: z.string().max(64).optional(),
  model: z.string().max(64).optional(),
  device: z.string().max(64).optional(),
  brand: z.string().max(64).optional(),
  hardware: z.string().max(64).optional(),
  buildFingerprint: z.string().max(200).optional(),
  buildId: z.string().max(64).optional(),
  /** "release-keys" | "test-keys" — test-keys は非公式 ROM のサイン */
  buildTags: z.string().max(64).nullable().optional(),
  buildType: z.string().max(32).optional(),

  // OS / リスク
  sdkInt: z.number().int().optional(),
  securityPatch: z.string().max(32).nullable().optional(),
  isDeviceSecure: z.boolean().optional(),
  adbEnabled: z.boolean().nullable().optional(),
  developmentSettings: z.boolean().nullable().optional(),
  isEmulator: z.boolean().optional(),

  // 環境
  timeZone: z.string().max(64).optional(),
  locale: z.string().max(32).optional(),
});

export type DeviceProfile = z.infer<typeof deviceProfileSchema>;

/** ラッパーが返す JSON: { profile: "<json 文字列>", signature: "<base64>" }。 */
const envelopeSchema = z.object({
  profile: z.string().min(2).max(4000),
  signature: z.string().min(1).max(400),
});

export type ProfileVerdict =
  | { ok: true; profile: DeviceProfile; payload: string; signature: string }
  | { ok: false };

/**
 * 署名済みプロファイルを検証する。
 *
 * 順序が重要: **署名検証 → parse → nonce 照合**。未検証の JSON を先に
 * parse しない。
 */
export function verifyDeviceProfile(
  publicKeySpkiB64: string,
  nonce: string,
  envelope: unknown,
): ProfileVerdict {
  const parsedEnvelope = envelopeSchema.safeParse(envelope);
  if (!parsedEnvelope.success) return { ok: false };
  const { profile: profileJson, signature } = parsedEnvelope.data;

  if (
    !verifyDeviceSignature(
      publicKeySpkiB64,
      attestPayload(nonce, profileJson),
      signature,
    )
  ) {
    return { ok: false };
  }

  let raw: unknown;
  try {
    raw = JSON.parse(profileJson);
  } catch {
    return { ok: false };
  }
  const parsed = deviceProfileSchema.safeParse(raw);
  if (!parsed.success) return { ok: false };
  // 別チャレンジで得た署名の貼り替えを防ぐ
  if (parsed.data.nonce !== nonce) return { ok: false };
  if (parsed.data.v !== PROFILE_SCHEMA_VERSION) return { ok: false };

  return {
    ok: true,
    profile: parsed.data,
    payload: profileJson,
    signature,
  };
}

/**
 * 検証済みプロファイル → 所有区分の判定材料。
 * `signatureVerified: true` は**検証を通ったここでしか作らない**
 * （型の上で、未検証のプロファイルが判定器に届かないようにするため）。
 */
export function toVerifiedWrapperProfile(
  profile: DeviceProfile,
): VerifiedWrapperProfile {
  return {
    signatureVerified: true,
    isDeviceOwner: profile.isDeviceOwner === true,
    isProfileOwner: profile.isProfileOwner === true,
    isManagedProfile: profile.isManagedProfile === true,
    enrollmentId: profile.enrollmentId ?? null,
    buildTags: profile.buildTags ?? null,
    adbEnabled: profile.adbEnabled ?? null,
    isEmulator: profile.isEmulator === true,
  };
}

File: external/android-kiosk/app/src/main/java/jp/co/ckk/kiosk/DeviceProfile.kt
package jp.co.ckk.kiosk

import android.app.KeyguardManager
import android.app.admin.DevicePolicyManager
import android.content.Context
import android.os.Build
import android.provider.Settings

/**
 * DeviceProfile — サーバーへ署名して渡す端末情報。
 *
 * 目的は「この端末が社給の管理端末か」をサーバーが判定できるようにすること。
 * 判定の根拠になるのは主に isDeviceOwner / isProfileOwner / enrollmentId で、
 * Build 情報とリスクフラグ（test-keys・ADB・エミュレータ）は SY09 の警告表示用。
 *
 * ■ JSON は **キー順を固定して手で組み立てる**
 * この文字列がそのまま署名対象になるので、JSONObject のイテレーション順に
 * 依存すると端末や OS 版の違いでサーバーが受け取る文字列が揺れる。
 * サーバー側の対向: coolify/apps/nextjs-kiosk/src/lib/device-profile.ts
 *
 * ■ 取れない項目は null
 * minSdk 29 なので enrollmentId（API31+）や installer（API30+）は実機でも
 * 普通に取れない。サーバーは null 前提で書いてある。
 */
object DeviceProfile {

    /** device-profile.ts の PROFILE_SCHEMA_VERSION と一致させること。 */
    private const val SCHEMA_VERSION = 1

    private fun dpm(context: Context): DevicePolicyManager =
        context.getSystemService(Context.DEVICE_POLICY_SERVICE) as DevicePolicyManager

    /** JSON 文字列としてエスケープ（制御文字も含めて安全側に倒す）。 */
    private fun esc(value: String): String {
        val sb = StringBuilder(value.length + 8)
        for (ch in value) {
            when {
                ch == '"' -> sb.append("\\\"")
                ch == '\\' -> sb.append("\\\\")
                ch == '\n' -> sb.append("\\n")
                ch == '\r' -> sb.append("\\r")
                ch == '\t' -> sb.append("\\t")
                ch < ' ' -> sb.append(String.format("\\u%04x", ch.code))
                else -> sb.append(ch)
            }
        }
        return sb.toString()
    }

    private class Json {
        private val sb = StringBuilder("{")
        private var first = true

        private fun comma() {
            if (first) first = false else sb.append(',')
        }

        fun str(key: String, value: String?): Json {
            comma()
            sb.append('"').append(key).append("\":")
            if (value == null) sb.append("null") else sb.append('"').append(esc(value)).append('"')
            return this
        }

        fun num(key: String, value: Long?): Json {
            comma()
            sb.append('"').append(key).append("\":").append(value?.toString() ?: "null")
            return this
        }

        fun bool(key: String, value: Boolean?): Json {
            comma()
            sb.append('"').append(key).append("\":").append(value?.toString() ?: "null")
            return this
        }

        fun build(): String = sb.append('}').toString()
    }

    private fun <T> orNull(block: () -> T): T? = try {
        block()
    } catch (_: Throwable) {
        // 権限不足・OEM 差異で落ちる呼び出しがある。1 項目のために
        // プロファイル全体を失わない。
        null
    }

    /** エミュレータ判定（完全ではない — 警告表示のための目安）。 */
    private fun isEmulator(): Boolean =
        Build.FINGERPRINT.startsWith("generic") ||
            Build.FINGERPRINT.contains("vbox") ||
            Build.FINGERPRINT.contains("emulator") ||
            Build.MODEL.contains("Emulator") ||
            Build.MODEL.contains("Android SDK built for") ||
            Build.PRODUCT == "google_sdk"

    /**
     * 署名対象の正規形 JSON を組み立てる。**キーの順序を変えないこと**
     * （順序を変えても署名は成立するが、差分を追う人が混乱する）。
     */
    fun build(context: Context, nonce: String): String {
        val dpm = orNull { dpm(context) }
        val admin = orNull { android.content.ComponentName(context, KioskDeviceAdminReceiver::class.java) }
        val deviceOwner = orNull { KioskMode.isDeviceOwner(context) } ?: false
        val profileOwner = orNull { dpm?.isProfileOwnerApp(context.packageName) } ?: false

        return Json()
            .num("v", SCHEMA_VERSION.toLong())
            .str("nonce", nonce)
            .num("signedAt", System.currentTimeMillis())

            .str("appVersion", BuildConfig.VERSION_NAME)
            .num("appVersionCode", BuildConfig.VERSION_CODE.toLong())
            .str("packageName", context.packageName)
            .str(
                "installer",
                orNull {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                        context.packageManager.getInstallSourceInfo(context.packageName)
                            .installingPackageName
                    } else {
                        @Suppress("DEPRECATION")
                        context.packageManager.getInstallerPackageName(context.packageName)
                    }
                },
            )

            // ── 管理状態（所有区分の判定材料） ──
            .bool("isDeviceOwner", deviceOwner)
            .bool("isProfileOwner", profileOwner)
            .bool(
                "isManagedProfile",
                orNull { admin?.let { a -> dpm?.isManagedProfile(a) } } ?: false,
            )
            .num(
                "activeAdmins",
                orNull { dpm?.activeAdmins?.size?.toLong() },
            )
            .num(
                "lockTaskState",
                orNull {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        (context.getSystemService(Context.ACTIVITY_SERVICE)
                            as android.app.ActivityManager).lockTaskModeState.toLong()
                    } else {
                        null
                    }
                },
            )
            // 組織 × 端末 × アプリで一意・初期化しても変わらない = 社給の最良の証拠。
            // device owner でないと取れない（API31+）。
            .str(
                "enrollmentId",
                orNull {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && deviceOwner) {
                        dpm?.enrollmentSpecificId?.takeIf { it.isNotEmpty() }
                    } else {
                        null
                    }
                },
            )

            // ── 端末同定 ──
            .str(
                "androidId",
                orNull {
                    Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
                },
            )
            .str(
                "serial",
                // device owner のときだけ取れる（それ以外は SecurityException）
                orNull { if (deviceOwner) Build.getSerial() else null },
            )

            // ── Build ──
            .str("manufacturer", Build.MANUFACTURER)
            .str("model", Build.MODEL)
            .str("device", Build.DEVICE)
            .str("brand", Build.BRAND)
            .str("hardware", Build.HARDWARE)
            .str("buildFingerprint", Build.FINGERPRINT)
            .str("buildId", Build.ID)
            // "test-keys" は非公式 ROM / root 化のサイン
            .str("buildTags", Build.TAGS)
            .str("buildType", Build.TYPE)

            // ── OS / リスク ──
            .num("sdkInt", Build.VERSION.SDK_INT.toLong())
            .str(
                "securityPatch",
                orNull {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        Build.VERSION.SECURITY_PATCH
                    } else {
                        null
                    }
                },
            )
            .bool(
                "isDeviceSecure",
                orNull {
                    (context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager)
                        .isDeviceSecure
                },
            )
            .bool(
                "adbEnabled",
                orNull {
                    Settings.Global.getInt(
                        context.contentResolver,
                        Settings.Global.ADB_ENABLED,
                        0,
                    ) == 1
                },
            )
            .bool(
                "developmentSettings",
                orNull {
                    Settings.Global.getInt(
                        context.contentResolver,
                        Settings.Global.DEVELOPMENT_SETTINGS_ENABLED,
                        0,
                    ) == 1
                },
            )
            .bool("isEmulator", isEmulator())

            // ── 環境 ──
            .str("timeZone", orNull { java.util.TimeZone.getDefault().id })
            .str("locale", orNull { java.util.Locale.getDefault().toLanguageTag() })
            .build()
    }
}
